### PR TITLE
Add 16-byte key-value support to `experimental::static_map`

### PR DESCRIPTION
**Pull request by PointKernel**
This PR adds `back_to_back_cas` and `cas_dependent_write` to the base `open_addressing_ref_impl` class. It allows the support of 8-byte keys and payloads in the new experimental map.

### CODE REVIEW

File: tests/static_map/unique_sequence_test.cu [hunk 1]
```diff
@@ -242,17 +242,31 @@ __inline__ void test_unique_sequence(Map& map, size_type num_keys)
...
+  (int64_t, int32_t, cuco::test::probe_sequence::double_hashing, 2),
+  (int64_t, int64_t, cuco::test::probe_sequence::double_hashing, 2),
+  (int32_t, int32_t, cuco::test::probe_sequence::linear_probing, 1),
+  (int32_t, int64_t, cuco::test::probe_sequence::linear_probing, 1),
+  (int32_t, int32_t, cuco::test::probe_sequence::linear_probing, 2),
+  (int32_t, int64_t, cuco::test::probe_sequence::linear_probing, 2),
+  (int64_t, int32_t, cuco::test::probe_sequence::linear_probing, 1),
+  (int64_t, int64_t, cuco::test::probe_sequence::linear_probing, 1),
+  (int64_t, int32_t, cuco::test::probe_sequence::linear_probing, 2),
+  (int64_t, int64_t, cuco::test::probe_sequence::linear_probing, 2))
```
Comment (PointKernel) [inline tests/static_map/unique_sequence_test.cu:278]: This is ugly, they must provide something like a parameter axis.

Comment (PointKernel) [reply tests/static_map/unique_sequence_test.cu:278]: Guess who complained about this first? https://github.com/catchorg/Catch2/issues/1989

File: include/cuco/detail/open_addressing_ref_impl.cuh [hunk 2]
```diff
@@ -555,53 +649,130 @@ class open_addressing_ref_impl {
...
+  {
+    auto const expected_key = this->empty_slot_sentinel_.first;
+
+    auto old_key = compare_and_swap(&slot->first, expected_key, value.first);
+
+    auto* old_key_ptr = reinterpret_cast<key_type*>(&old_key);
+
+    // if key success
+    if (cuco::detail::bitwise_compare(*old_key_ptr, expected_key)) {
+      atomic_store(&slot->second, value.second);
```
Comment (PointKernel) [inline include/cuco/detail/open_addressing_ref_impl.cuh:737]: Actually, can this just be a plain assignment?
```suggestion
      slot->second = value.second;
```

Comment (PointKernel) [reply include/cuco/detail/open_addressing_ref_impl.cuh:737]: Leave it as `atomic_store` for now for simplicity.

File: include/cuco/detail/open_addressing_ref_impl.cuh [hunk 3]
```diff
@@ -253,7 +256,13 @@ class open_addressing_ref_impl {
         // If the key is already in the container, return false
         if (eq_res == detail::equal_result::EQUAL) { return {iterator{&window_ptr[i]}, false}; }
         if (eq_res == detail::equal_result::EMPTY) {
-          switch (attempt_insert(window_ptr + i, value, predicate)) {
+          switch ([&]() {
+            if constexpr (sizeof(value_type) <= 8) {
+              return packed_cas(window_ptr + i, value, predicate);
+            } else {
+              return cas_dependent_write(window_ptr + i, value, predicate);
```
Comment (sleeepyjack) [inline include/cuco/detail/open_addressing_ref_impl.cuh:263]: I wonder if this is something we schould make configurable from the outside once adapt the mixin stuff.

File: include/cuco/detail/open_addressing_ref_impl.cuh [hunk 4]
```diff
@@ -253,7 +256,13 @@ class open_addressing_ref_impl {
         // If the key is already in the container, return false
         if (eq_res == detail::equal_result::EQUAL) { return {iterator{&window_ptr[i]}, false}; }
         if (eq_res == detail::equal_result::EMPTY) {
-          switch (attempt_insert(window_ptr + i, value, predicate)) {
+          switch ([&]() {
```
Comment (sleeepyjack) [inline include/cuco/detail/open_addressing_ref_impl.cuh:259]: This notation is wild! I love it.

File: include/cuco/detail/open_addressing_ref_impl.cuh [hunk 5]
```diff
@@ -541,10 +555,92 @@ class open_addressing_ref_impl {
...
+  __device__ constexpr auto compare_and_swap(T* address, T expected, T desired)
+  {
+    // temporary workaround due to performance regression
+    // https://github.com/NVIDIA/libcudacxx/issues/366
+    if constexpr (sizeof(T) == sizeof(unsigned int)) {
+      auto* expected_ptr = reinterpret_cast<unsigned int*>(&expected);
+      auto* desired_ptr  = reinterpret_cast<unsigned int*>(&desired);
+      if constexpr (Scope == cuda::thread_scope_system) {
+        return atomicCAS_system(
+          reinterpret_cast<unsigned int*>(address), *expected_ptr, *desired_ptr);
```
Comment (sleeepyjack) [inline include/cuco/detail/open_addressing_ref_impl.cuh]:
```suggestion
      auto* const slot_ptr     = reinterpret_cast<unsigned int*>(address);
      auto const * const expected_ptr = reinterpret_cast<unsigned int*>(&expected);
      auto const * const desired_ptr  = reinterpret_cast<unsigned int*>(&desired);
      if constexpr (Scope == cuda::thread_scope_system) {
        return atomicCAS_system(
          slot_ptr, *expected_ptr, *desired_ptr);
```

Minor. Although I'm not sure about the proposed const-correctness changes.

**Review by sleeepyjack — APPROVED** [review]
LGTM.
I wonder, is there any notable perf difference between the old and new `static_map` implementation?

**Pull request merged**